### PR TITLE
Attributes indexes 1 by 1

### PR DESCRIPTION
**Pull request by fogelito**
This PR on createCollection will create the attributes and indexes 1 by 1 not by schema.

### CODE REVIEW

File: src/Database/Database.php [hunk 1]
```diff
@@ -13,6 +13,7 @@
 use Utopia\Database\Exception\Limit as LimitException;
 use Utopia\Database\Exception\Structure as StructureException;
 use Utopia\Cache\Cache;
+use function PHPUnit\Framework\throwException;
```
Comment (TorstenDittmann) [inline src/Database/Database.php:16]: unused import?

File: src/Database/Database.php [hunk 2]
```diff
@@ -531,33 +527,73 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+        $this->createCollectionIndexes($id, $indexes);
 
-            if (
-                $this->adapter->getRowLimit() > 0 &&
-                $this->adapter->getAttributeWidth($collection) > $this->adapter->getRowLimit()
-            ) {
-                throw new LimitException('Row width limit of ' . $this->adapter->getRowLimit() . ' exceeded. Cannot create collection.');
-            }
+        $collection->setAttribute('attributes', $attributes);
+        $collection->setAttribute('indexes', $indexes);
```
Comment (TorstenDittmann) [inline src/Database/Database.php:540]:
```suggestion
        $collection
            ->setAttribute('attributes', $attributes);
            ->setAttribute('indexes', $indexes);
```

File: src/Database/Database.php [hunk 3]
```diff
@@ -531,33 +527,73 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+     *
+     * @param string $collection
+     * @param Document[] $attributes
+     *
+     * @return bool
+     */
+    public function createCollectionAttributes(string $collection, array $attributes = []): bool
+    {
+        foreach ($attributes as $attribute){
+            self::createAttribute(
```
Comment (TorstenDittmann) [inline src/Database/Database.php:557]:
```suggestion
            $this->createAttribute(
```

File: src/Database/Database.php [hunk 4]
```diff
@@ -531,33 +527,73 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+                $collection,
+                $attribute->getId(),
+                $attribute->getAttribute('type'),
+                $attribute->getAttribute('size') ?? 0,
+                $attribute->getAttribute('required'),
+                $attribute->getAttribute('default'),
+                $attribute->getAttribute('signed') ?? false,
+                $attribute->getAttribute('array') ?? false,
+                $attribute->getAttribute('format'),
+                $attribute->getAttribute('formatOptions')?? [],
```
Comment (TorstenDittmann) [inline src/Database/Database.php:567]:
```suggestion
                $attribute->getAttribute('formatOptions') ?? [],
```

File: src/Database/Database.php [hunk 5]
```diff
@@ -531,33 +527,73 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+     * @param string $collection
+     * @param Document[] $indexes
+     *
+     * @return bool
+     * @throws Exception
+     */
+    public function createCollectionIndexes(string $collection, array $indexes = []): bool
+    {
+        foreach ($indexes as $index){
+            self::createIndex(
```
Comment (TorstenDittmann) [inline src/Database/Database.php:586]:
```suggestion
            $this->createIndex(
```

File: src/Database/Database.php [hunk 6]
```diff
@@ -531,33 +527,73 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+    /**
+     * Create Collection Attributes
+     *
+     * @param string $collection
+     * @param Document[] $indexes
+     *
+     * @return bool
+     * @throws Exception
+     */
+    public function createCollectionIndexes(string $collection, array $indexes = []): bool
```
Comment (TorstenDittmann) [inline src/Database/Database.php:583]:
```suggestion
    public function createIndexes(string $collection, array $indexes = []): bool
```

File: src/Database/Database.php [hunk 7]
```diff
@@ -531,33 +527,73 @@ public function createCollection(string $id, array $attributes = [], array $inde
...
+
+    /**
+     * Create Collection Attributes
+     *
+     * @param string $collection
+     * @param Document[] $attributes
+     *
+     * @return bool
+     */
+    public function createCollectionAttributes(string $collection, array $attributes = []): bool
```
Comment (TorstenDittmann) [inline src/Database/Database.php:554]:
```suggestion
    public function createAttributes(string $collection, array $attributes = []): bool
```

File: src/Database/Database.php [hunk 8]
```diff
@@ -618,25 +654,35 @@ public function deleteCollection(string $id): bool
...
      *
      * @return bool
+     * @throws DuplicateException
+     * @throws LimitException
      */
-    public function createAttribute(string $collection, string $id, string $type, int $size, bool $required, $default = null, bool $signed = true, bool $array = false, string $format = null, array $formatOptions = [], array $filters = []): bool
+    public function createAttribute(string $collection, string $id, string $type, int $size, bool $required, $default = null, bool $signed = false, bool $array = false, string $format = null, array $formatOptions = [], array $filters = []): bool
     {
+        if(empty($collection)){
+            throw new Exception('CreateAttribute error empty collection');
```
Comment (TorstenDittmann) [inline src/Database/Database.php:671]:
```suggestion
            throw new Exception("Collection can't be empty.");
```

File: src/Database/Database.php [hunk 9]
```diff
@@ -618,25 +654,35 @@ public function deleteCollection(string $id): bool
...
+        }
+
+        if(empty($id)){
+            throw new Exception('CreateAttribute error empty id');
```
Comment (TorstenDittmann) [inline src/Database/Database.php:675]:
```suggestion
            throw new Exception("ID can't be empty");
```

File: src/Database/Database.php [hunk 10]
```diff
@@ -618,25 +654,35 @@ public function deleteCollection(string $id): bool
...
+
         $collection = $this->getCollection($collection);
 
         // attribute IDs are case insensitive
         $attributes = $collection->getAttribute('attributes', []);
         /** @var Document[] $attributes */
         foreach ($attributes as $attribute) {
-            if (\strtolower($attribute->getId()) === \strtolower($id)) {
-                throw new DuplicateException('Attribute already exists');
+            if (\strtolower($attribute->getId()) === \strtolower($id) && $collection->getId() != '_metadata') {
```
Comment (TorstenDittmann) [inline src/Database/Database.php:684]:
```suggestion
            if (\strtolower($attribute->getId()) === \strtolower($id) && $collection->getId() !== self::METADATA) {
```

**Review by TorstenDittmann — CHANGES_REQUESTED** [review]

File: src/Database/Adapter/MariaDB.php [hunk 11]
```diff
@@ -123,86 +123,42 @@ public function delete(string $name): bool
...
+            CREATE TABLE IF NOT EXISTS `{$database}`.`{$namespace}_{$id}_perms` (
+            `_id` int(11) unsigned NOT NULL AUTO_INCREMENT,
+            `_type` VARCHAR(12) NOT NULL,
+            `_permission` VARCHAR(255) NOT NULL,
+            `_document` VARCHAR(255) NOT NULL,
+            PRIMARY KEY (`_id`),
+            UNIQUE INDEX `_index1` (`_document`,`_type`,`_permission`),
+            INDEX `_index2` (`_permission`)
+        )")
+        ->execute();
```
[the remark this answers is not in the document]
Comment (fogelito) [reply src/Database/Adapter/MariaDB.php:160]: OK

File: src/Database/Database.php (hunk 1 above)
Comment (fogelito) [reply src/Database/Database.php:16]: REMOVED

File: src/Database/Database.php (hunk 2 above)
Comment (fogelito) [reply src/Database/Database.php:540]: CHANGED

**Comment by eldadfux** [comment]
We decided to not continue with this route due to performance issues. Creating an index one by one turned out to be significantly slower.

**Pull request closed without merging**